nni/contrib/distillation/storage.py

```diff
@@ -0,0 +1,260 @@
+        root_path.mkdir(parents=True, exist_ok=True)
+        uid2filename = {}
+        for uid, data in self._uid2data.items():
+            uid2filename[uid] = f'{uid}.pkl'
+            pickle_dump(data, root_path / 'data' / uid2filename[uid])
+        with Path(root_path / SETTING_FILENAME).open(mode='w') as f:
+            json_tricks.dump(uid2filename, f, indent=4)
+
+    @classmethod
+    def from_checkpoint(cls, checkpoint_folder: str | PathLike):
+        root_path = Path(checkpoint_folder).absolute()
+        assert root_path.exists()
+        # load uid2filename dict
+        with Path(root_path / SETTING_FILENAME).open(mode='r') as f:
+            uid2filename = json_tricks.load(f)
+        # load data
+        uid2data = {}
+        for uid, filename in uid2filename.items():
+            uid2data[uid] = pickle_load(root_path / 'data' / filename)
+
+        storage = cls()
+        storage._uid2data.update(uid2data)
+        return storage
+
+
+class FileStorage(_DistilStorage):
+    """
+    Cache the distillation labels on disk.
+    Under each label will be saved as a file named `{uid}.pkl` under `cache_folder`.
+    """
+    def __init__(self, cache_folder: str | PathLike):
+        self._cache_folder = Path(cache_folder).absolute()
+        self._data_folder = self._cache_folder / 'data'
+        self._data_folder.mkdir(parents=True, exist_ok=True)
+        self._uid2filename: Dict[str, str] = {}
+
+    def __len__(self):
+        return len(self._uid2filename)
+
+    def record(self, uid: str, data: Any):
+        if uid in self._uid2filename:
+            warn_msg = f'Already record distil data for {uid}, update distil data for uid {uid}.'
+            _logger.warning(warn_msg)
+        self._uid2filename[uid] = f'{uid}.pkl'
+        pickle_dump(data, self._data_folder / self._uid2filename[uid])
+
+    def select(self, uid: str) -> Any:
+        return pickle_load(self._data_folder / self._uid2filename[uid])
+
+    def get_uids(self) -> List[str]:
+        uids = list(self._uid2filename.keys())
+        uids.sort()
+        return uids
+
+    def save(self, checkpoint_folder: str | PathLike):
+        root_path = Path(checkpoint_folder).absolute()
+        if root_path != self._cache_folder:
+            shutil.copytree(self._cache_folder, root_path, copy_function=shutil.copy2, dirs_exist_ok=False)
+        with Path(root_path / SETTING_FILENAME).open(mode='w') as f:
+            json_tricks.dump(self._uid2filename, f, indent=4)
+
+    def cleanup(self):
+        """
+        Cleanup the saved files under `cache_folder`.
+        """
+        cleanup_filename = set(self._uid2filename.values())
+        if Path(self._cache_folder / SETTING_FILENAME).exists():
+            with Path(self._cache_folder / SETTING_FILENAME).open(mode='r') as f:
+                uid2filename: Dict[str, str] = json_tricks.load(f)
+                cleanup_filename = cleanup_filename.difference(uid2filename.values())
```

Comment:
according to this description, cleanup function is used to cleanup the saved files under cache_folder,  why dose use cleanup_filename,difference() to find cleanup_filename?  When will cleanup_filename and uid2filename be different?

Reply:
it's a good question. you can find `if Path(self._cache_folder / SETTING_FILENAME).exists()` is True, that means this is a checkpoint folder, and some filenames are loaded from the `Path(self._cache_folder / SETTING_FILENAME)`, then when cleanup, we should only delete the files generated by this storage, not the files loaded from the checkpoint.